src/Chapter3.hs

```diff
@@ -910,7 +983,21 @@ Implement instances of "Append" for the following types:
 class Append a where
     append :: a -> a -> a
 
+newtype Gold = Gold Int deriving(Show)
+
+instance Append Gold where
+  append :: Gold -> Gold -> Gold
+  append (Gold a) (Gold b) = Gold $ a + b
+
+instance Append [a] where
+  append :: [a] -> [a] -> [a]
+  append xs ys = xs ++ ys
 
+instance Append a => Append (Maybe a) where
+  append :: Maybe a -> Maybe a -> Maybe a
+  append Nothing _ = Nothing
+  append _ Nothing = Nothing
```

Comment:
While this is valid implementation, in reality, it would be more beneficial is this function returned `Just` when at least one argument is `Just`. And this is actually how Maybe works on a similar existing typeclass in Haskell 🙂

Reply:
do you mean something like that?
```haskell
append Nothing Nothing = Nothing
append a Nothing = a
append Nothing b = b
```

Reply:
Yes, this is an absolutely correct solution!

However, I can notice that these 3 cases with `Nothing`s could be reduced to 2!
Let's see what we can unite in here.

If some of the arguments are `Nothing`, you actually always return the second one! So you need to inspect only two cases: `Nothing` in the first position, and `Nothing` on the second one 🙂

Reply:
oooh, so in case of two `Nothing`s `x` will match to `Nothing` and we are returning `x`
I didnt think about it, thats smart 🙂